pkg/scaffold/internal/util/path.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/gobuffalo/flect"
+
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/resource"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/webhook"
+)
+
+const namedParameterPattern = "%\\[[\\w_-]+\\]s"
+
+var namedParameterRegexp = regexp.MustCompile(namedParameterPattern)
+
+// sprintm replaces any %[KEY]s contained in the format string with the value in
+// the parameters map for KEY. If not found, it leaves it as it was.
+func sprintm(format string, parameters map[string]string) string {
+	return namedParameterRegexp.ReplaceAllStringFunc(format, func(key string) string {
+		if value, found := parameters[key[2:len(key)-2]]; found {
+			return value
+		}
+		return key
+	})
+}
+
+func Path(path string, parameters ...interface{}) string {
```

Comment:
Looking at the invocation of `util.Path`, I was wondering if you considered this API:
```Go
func PathJoin(paths ...string, params interface{}) string
```

Where Path can internally call `filepath.Join(paths...)` followed by named param substitution. I think this can help with the code readability where `util.Path` is being called.

Reply:
Thats sounds interesting. Can you provide arguments after a `...string`?

If that is not allowed we could do something like:

```go
func JoinPath(paths ...string) Pather {
    return Pather{filepath.Join(paths...)}
}

type Pather struct {
    path string
}

func (p Pather) Replace(params map[string]string) string {...}
```

It would be called like `PathJoin(paths...).Replace(params)`. We could also go for the other way round, creating a type that holds the `params` and provides a `Join` method: `Formatter(params).Join(paths...)` or the simplest option: `PathJoin(params, paths...)`.

Reply:
>Can you provide arguments after a ...string?

No, we can't :). 

I like all the suggestions you listed. Don't have strong opinion about any particular one. I looked at https://golang.org/pkg/strings/#Replacer for some inspiration and your suggestion for `Formatter(params).Join(paths...)` can be tweaked a bit where `resource` can implement method called `Replacer()` which returns a replacer implementation that can be invoked with `Replace()` or `Join` (in our case). So it will look like:

```Go
# assuming r points to API resource object and APIResource implements Replacer() method
r := APIResource{}

# join path version
r.Replacer().Join(paths....)

# normal string replace if we need in future.
r.Replacer().Replace(s)

```

Reply:
Okey, I do like the design. Adding a `Replacer` method to `pkg/model/resource.Resource` that provides those two methods, `Join(...string) string` and `Replace(string) string`.

The `%[KEY]s` syntax is okey for the replacement? We would also have to revisit which fields should be replaced, for example, should `kind` key refer to lowercase Kind or should we use `lower(kind)`, `lower-kind` or `kind-lower`? I will try to get this design implemented when I get some time and that way I will also build a list of which items we want to be replaced so that we can find which keys we think fit best.

Reply:
%[KEY]s seems fine to me. I think we should start with lower-case by default (don't expose lower/upper directives yet) and only expose those if needed in the future because I doubt we need those.

Reply:
@droot I guessed that, if we were gonna follow `strings.Replacer` format, we may also use it directly. Let me know what you think about this implementation.